stdlib/public/core/BinaryFloatingPoint.swift

```diff
@@ -0,0 +1,677 @@
+  ///     initializer has the same sign as `signOf`.
+  ///   - magnitudeOf: A value from which to use the magnitude. The result of
+  ///     the initializer has the same magnitude as `magnitudeOf`.
+  @inlinable // FIXME(sil-serialize-all)
+  public init(signOf: Self, magnitudeOf: Self) {
+    self.init(sign: signOf.sign,
+              exponentBitPattern: magnitudeOf.exponentBitPattern,
+              significandBitPattern: magnitudeOf.significandBitPattern)
+  }
+  
+  @inlinable // FIXME(sil-serialize-all)
+  public // @testable
+  static func _convert<Source : BinaryInteger>(
+    from source: Source
+    ) -> (value: Self, exact: Bool) {
+    guard _fastPath(source != 0) else { return (0, true) }
+    
+    let magnitude = source.magnitude
+    
+    let exponent = Int(magnitude._binaryLogarithm())
+    let exemplar = Self.greatestFiniteMagnitude
+    guard _fastPath(exponent <= exemplar.exponent) else {
+      return Source.isSigned && source < 0
+        ? (-.infinity, false)
+        : (.infinity, false)
+    }
+    let exponentBitPattern =
+      (1 as Self).exponentBitPattern /* (i.e., bias) */
+        + Self.RawExponent(exponent)
+    
+    let maxSignificandWidth = exemplar.significandWidth
+    let shift = maxSignificandWidth &- exponent
+    let significandBitPattern = shift >= 0
+      ? Self.RawSignificand(magnitude) << shift & exemplar.significandBitPattern
+      : Self.RawSignificand(
+        magnitude >> -shift & Source.Magnitude(exemplar.significandBitPattern))
+    
+    let value = Self(
+      sign: Source.isSigned && source < 0 ? .minus : .plus,
+      exponentBitPattern: exponentBitPattern,
+      significandBitPattern: significandBitPattern)
+    
+    if exponent &- magnitude.trailingZeroBitCount <= maxSignificandWidth {
+      return (value, true)
+    }
+    // We promise to round to the closest representation, and if two
+    // representable values are equally close, the value with more trailing
+    // zeros in its significand bit pattern. Therefore, we must take a look at
+    // the bits that we've just truncated.
+    let ulp = (1 as Source.Magnitude) << -shift
+    let truncatedBits = magnitude & (ulp - 1)
+    if truncatedBits < ulp / 2 {
+      return (value, false)
+    }
+    let rounded = Source.isSigned && source < 0 ? value.nextDown : value.nextUp
+    guard _fastPath(
+      truncatedBits != ulp / 2 ||
+        exponentBitPattern.trailingZeroBitCount <
+        rounded.exponentBitPattern.trailingZeroBitCount) else {
+          return (value, false)
+    }
+    return (rounded, false)
+  }
+  
+  /// Creates a new value, rounded to the closest possible representation.
+  ///
+  /// If two representable values are equally close, the result is the value
+  /// with more trailing zeros in its significand bit pattern.
+  ///
+  /// - Parameter value: The integer to convert to a floating-point value.
+  @inlinable // FIXME(sil-serialize-all)
```

Comment:
You should probably upgrade this to `@_transparent` or at least `@inline(__always)`. This *definitely* has to get compiled down to a single instruction for common types, even in -Onone builds, right?

Reply:
Yeah, there's some follow-on work for the concrete types here, which is why this is [nomerge]. Just making sure that nothing's broken with the first step of this change.

Reply:
... and actually there's a bunch of cases where we should be able to generate single-instruction conversions in generic contexts but fail to do so today even with optimization enabled, so we clearly need some more test coverage here.

Reply:
@jrose-apple I don't think we always inline at -Onone. If my memory is correct, this is something that I think would be good to put in.